rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,198 @@
+
+## Proposal
+
+A backfill mechanism can simply be defined as an algorithm for fetching,
+verifying and storing, headers and validator sets of a height prior to the
+current base of the node's blockchain. In matching the terminology used for
+other data retrieving protocols (i.e. fast sync and state sync), we
+call this method **ReverseSync**.
+
+We will define the mechanism in four sections:
+- Usage
+- Verification
+- Design
+- Termination
+
+### Usage
+
+For now, we focus purely on the case of a state syncing node, whom after
+syncing to a height will need to verify an adequate amount of previous data in
+order to be capable of processing new blocks and advancing sate. We can call
+this the `max_historical_height`/`time` (max as in earliest/highest possible
+time/height). Both height and time are necessary as this maps to the BFT time
+used for evidence expiration. After acquiring `State`, we calculate these
+parameters as:
+
+```go
+max_historical_height = max(state.InitialHeight, state.LastBlockHeight - state.ConsensusParams.EvidenceAgeHeight)
+max_historical_time = max(GenesisTime, state.LastBlockTime.Sub(state.ConsensusParams.EvidenceAgeTime))
+```
+
+Before starting either fast sync or consensus, we then run the following
+synchronous process:
+
+```go
+func ReverseSync(
+  max_historical_height int64
+  max_historical_time time.Time
+) error
+```
+
+Where we fetch and verify blocks until a block `A` where
+`A.Height <= max_historical_height` and `A.Time <= max_historical_time`.
+
+Upon successfully reverse syncing a node can now safely continue. As this
+feature is only used as part of state sync, one can think of this as merely an
+extension to it.
+
+### Verification
+
+ReverseSync is used to fetch and verify the following data structures:
+- `Header`
+- `ValidatorSet`
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. It then fetches the validator set
+and validates it as such:
+
+```go
+header[height-1].ValidatorHash == hash(validatorSet[height-1])
+```
+
+### Design
+
+This section will provide a high level overview of some of the more important
+characteristics of the design, saving the more tedious details as an ADR. This
+section is divided into two approaches that can be used to achieve the defined
+backfill mechanism based on the two networking protocols: RPC or P2P.
+
+#### RPC
+
+The aforementioned data is already available via the following RPC endpoints:
+`/blockchain` or `/commit` for `Header`'s' and `/validators` for
+`ValidatorSet`'s'. It may be the best option to fetch and verify these resources
+over RPC. Statesync already requires a list of RPC endpoints, so these could be
```

Comment:
I think we should really consider Erik's light client reactor

Reply:
Yeah if we went with extending the p2p layer then I think Erik's approach would be the strongest candidate when it comes to the actual implementation.

Reply:
There's this closed PR https://github.com/tendermint/tendermint/pull/4508

Reply:
That's about all the documentation we have on that